### PR TITLE
Upgrade dependencies

### PR DESCRIPTION
**Pull request by Veetaha**
It's been a long time, and this crate still serves us a good job. Time to migrate it to the latest world version =)

Maybe we should make it `1.0` already? The interface seems very stable.

### CODE REVIEW

File: Cargo.toml
Comment (Veetaha) [inline Cargo.toml:6]: Removed doc links. IIRC docs.rs is assumed by default, besides that one of them was already out of date referencing `0.4` version instead

**Comment by Veetaha** [comment]
Kindly ping @TedDriggs. Looks like you may be short on time maintaining this project? I'd be grateful if you could share the permissions for maintaining this repo and release the crate to crates.io with me (there is an option to invite in crate's page `Settings` tab). We are using it in [our company ](https://github.com/elastio), and I have made many contributions to OSS (rust-analyzer, dynomite), though I am more focused on our startup recently, but I still would like to at least periodically keep our dependencies up-to-date.

**Comment by Veetaha** [comment]
The repo will stay under your github user, and I won't claim any ownership, I am going to respect `MIT/Apache-2.0` license as it is defined in `Cargo.toml`

**Comment by TedDriggs** [comment]
What problem does this PR solve? It appears to increase the MSRV, which would mean future updates would need to target both version lineages, and the edition of this crate doesn’t impact its API.

Given the recent uptick in supply-chain attacks, I am not adding new maintainers to my projects.

I will continue to review PRs though.

**Comment by Veetaha** [comment]
Well, if you insist on staying with edition 2018, I'll leave it there, otherwise, it's just a dependency version bump.
The problem it solves it `cargo-deny` lint in our repo that reports that we have both `darling 0.10` and `darling 0.13` compiled into our binary, because crates with `0` major version and different minor versions are considered incompatible according to Cargo's semver rules.

**Comment by TedDriggs** [comment]
Thanks for clarifying; that is a great reason to merge this. I’ll take a look at it on some local tests today and merge soon.

**Comment by Veetaha** [comment]
Thanks, the I've reverted the edition change

**Review by TedDriggs — CHANGES_REQUESTED** [review]
Apologies for the slow review on this; life got very busy. To avoid any risk of a breaking change, I'd suggest we jump to 0.7.0 or 1.0.0

**Comment by Veetaha** [comment]
@TedDriggs Sure, let's give it `1.0`

File: from_variants_impl/src/from_impl.rs
Comment (TedDriggs) [inline from_variants_impl/src/from_impl.rs]: I had to put spaces in here to avoid `>>` being interpreted as one token, breaking the output completely. Has that now been fixed?

Comment (Veetaha) [reply from_variants_impl/src/from_impl.rs]: Was that ever a bug? I saw that `rustfmt::skip` and wondered why it was there and just saw how it only reformats the `>>`.
I don't think such bug exists, Rust's syntax clearly doesn't allow `>>` operator in type context, the lexer should never have a problem figuring this out. It may be a problem in C++ templates where type/value context can be confused by the compiler, but such problem doesn't exist in Rust.

Comment (TedDriggs) [reply from_variants_impl/src/from_impl.rs]: I need to try this when I get home. I believe the issue is that quasi-quoting doesn't run in type context, so if it infers that you're trying to push the "shift-left" symbol that's the Token emitted, and that is incorrect.

Comment (Veetaha) [reply from_variants_impl/src/from_impl.rs]: Hmm, that would be a pretty disruptive bug. I am pretty sure I've been spelling types like `Option<Vec<_>>` with `syn::parse_quote!()` and `quote::quote!()` and never had this problem. I've just rerun the `cargo test` and it worked for me.

I think rustc and `quote/syn` are smart enough to take apart `shift-right` operator from an angle bracket. IIRC, `rustc` only verifies that all `([{` have a matching and balanced  `}])`, and maybe several more simple invariants. 

According to the Rust docs of `proc-macro` crate for [`Punct` type](https://doc.rust-lang.org/stable/proc_macro/struct.Punct.html):
> Multi-character operators like += are represented as two instances of Punct with different forms of Spacing returned.

So, the compiler itself doesn't decide what `>>/<<` mean in the macro callsite. It passes that to the proc macro the following way. I tested it with this:

```rust
fn main() {
    proc_macro_joint::debug_log_tokens!(
        1 >> 2
    );

    proc_macro_joint::debug_log_tokens!(
        <<Option<Vec<String>>>>::None
    );
}
```

```rust
#[proc_macro]
pub fn debug_log_tokens(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    panic!("Input: {input:#?}");
}
```

The compiler output is the following

<details>
<summary>Compile output</summary>

```
error: proc macro panicked
 --> crates/proc-macro-joint-main/src/main.rs:5:5
  |
5 | /     proc_macro_joint::debug_log_tokens!(
6 | |         1 >> 2
7 | |     );
  | |_____^
  |
  = help: message: Input: TokenStream [
              Literal {
                  kind: Integer,
                  symbol: "1",
                  suffix: None,
                  span: #0 bytes(64..65),
              },
              Punct {
                  ch: '>',
                  spacing: Joint,
                  span: #0 bytes(66..68),
              },
              Punct {
                  ch: '>',
                  spacing: Alone,
                  span: #0 bytes(66..68),
              },
              Literal {
                  kind: Integer,
                  symbol: "2",
                  suffix: None,
                  span: #0 bytes(69..70),
              },
          ]

error: proc macro panicked
  --> crates/proc-macro-joint-main/src/main.rs:9:5
   |
9  | /     proc_macro_joint::debug_log_tokens!(
10 | |         <<Option<Vec<String>>>>::None
11 | |     );
   | |_____^
   |
   = help: message: Input: TokenStream [
               Punct {
                   ch: '<',
                   spacing: Joint,
                   span: #0 bytes(128..130),
               },
               Punct {
                   ch: '<',
                   spacing: Alone,
                   span: #0 bytes(128..130),
               },
               Ident {
                   ident: "Option",
                   span: #0 bytes(130..136),
               },
               Punct {
                   ch: '<',
                   spacing: Alone,
                   span: #0 bytes(136..137),
               },
               Ident {
                   ident: "Vec",
                   span: #0 bytes(137..140),
               },
               Punct {
                   ch: '<',
                   spacing: Alone,
                   span: #0 bytes(140..141),
               },
               Ident {
                   ident: "String",
                   span: #0 bytes(141..147),
               },
               Punct {
                   ch: '>',
                   spacing: Joint,
                   span: #0 bytes(147..149),
               },
               Punct {
                   ch: '>',
                   spacing: Joint,
                   span: #0 bytes(147..149),
               },
               Punct {
                   ch: '>',
                   spacing: Joint,
                   span: #0 bytes(149..151),
               },
               Punct {
                   ch: '>',
                   spacing: Joint,
                   span: #0 bytes(149..151),
               },
               Punct {
                   ch: ':',
                   spacing: Joint,
                   span: #0 bytes(151..153),
               },
               Punct {
                   ch: ':',
                   spacing: Alone,
                   span: #0 bytes(151..153),
               },
               Ident {
                   ident: "None",
                   span: #0 bytes(153..157),
               },
           ]
````
</details>

This proves that `rustc` doesn't do anything fancy to analyze the semantics of angle brackets, it just leaves this for the proc macro implementation, giving it [Spacing ](https://doc.rust-lang.org/stable/proc_macro/enum.Spacing.html) information.

Then `syn::parse_quote!()` is intelligent enough to understand what `>>` means, because it has the context of the expected syn AST node type, so knows what to expect.
And regarding `quote::quote!()`: it is very dumb, it doesn't assign any semantic meaning to `>>`, it just converts each input token to `proc_macro::TokenTree` using its Spacing API. For example, for multi-symbol tokens it [does this](https://github.com/dtolnay/quote/blob/278057d0d2e59c7de95d79d42e4934783d013e83/src/runtime.rs#L306-L336). And once `quote!()` forwards the raw `TokenTree` with the spacing properly set up, `rustc` analyzes the output of the proc macro as regular Rust code, so it compiles just fine

Comment (Veetaha) [reply from_variants_impl/src/from_impl.rs]: Sorry for verbosity. I always try to fight the "I don't know how, but it works, and we should better not touch it. Ever"

Comment (TedDriggs) [reply from_variants_impl/src/from_impl.rs]: Thanks for putting the effort into verifying this isn't an issue! I agree with your philosophy of questioning these sorts of limitations; I've now re-enabled the CI workflow, so hopefully one more push will be all that's needed to get the tests to run and then we'll be good to go.

Really appreciate your diligence in running this down.

File: from_variants_impl/Cargo.toml
Comment (TedDriggs) [inline from_variants_impl/Cargo.toml:13]: Should we bump the rest of these to their latest versions?

**Review by TedDriggs — APPROVED** [review]
Approved pending the CI run passing.

File: from_variants_impl/Cargo.toml
Comment (Veetaha) [reply from_variants_impl/Cargo.toml:13]: They are already at their latest versions. `proc-macro2/syn/quote` are developed by David Tolnay. And he makes his crates extra-stable. He knows how much burden it is for the ecosystem to keep up-to-date. That's why he always does only patch releases.
You can see for yourself:
- [`proc-macro2` crate](https://docs.rs/proc-macro2/latest/proc_macro2/): `1.0.36`
- [`quote` crate](https://docs.rs/quote/latest/quote/): `1.0.15`
- [`syn` crate](https://docs.rs/syn/latest/syn/): `1.0.86`

**Comment by TedDriggs** [comment]
Can you do a no-op push to get CI to run? If GH allows me to force-start a CI workflow on a pull request, I can’t find the option

**Comment by Veetaha** [comment]
Hmm, I am looking into the test failure... For some reason it passes in my env.

**Comment by Veetaha** [comment]
Looks like spacing does make a difference in that test there. It appears I was running the tests with simple `cargo test`, so not all of them were run, because this cargo workspace is non-uniform. There is a default package in this workspace, so `cargo test` commands without explicit `--all/--workspace` operate only on a single crate. I am just used to cargo workspaces, where there isn't a default package, so I've missed this.
So the problem `rustfmt::skip` was solving is not the compilation error, but the problem that when comparing token streams, spacing in their tokens matters.
I've returned `rustfmt::skip` with more context in the comment

**Comment by Veetaha** [comment]
@TedDriggs now it should be ready to merge

**Comment by TedDriggs** [comment]
I will publish tomorrow! Thanks so much for doing this

**Pull request merged**
